File: pc_manager/src/tasks/reboot.c
#include "../conn.h"

#if __linux

#  include <unistd.h>
#  include <linux/reboot.h>
#  include <syscall.h>
#  include <stdbool.h>

#elif WIN32
#  include <Windows.h>
#  define bool BOOL
#  define true TRUE
#  define false FALSE
#endif

int task_reboot() {
#if __linux
    if (syscall(SYS_reboot, LINUX_REBOOT_MAGIC1, LINUX_REBOOT_MAGIC2, LINUX_REBOOT_CMD_RESTART, 0) == -1) {
        perror("Failed with: ");
        return -1;
    }
#elif WIN32
    HANDLE hToken;
    TOKEN_PRIVILEGES tkp;

    // Get a token for this process.
    if (!OpenProcessToken(GetCurrentProcess(),
                          TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY, &hToken)) {
        return -1;
    }

    // Get the LUID for the shutdown privilege.
    LookupPrivilegeValue(NULL, SE_SHUTDOWN_NAME,
                         &tkp.Privileges[0].Luid);

    tkp.PrivilegeCount = 1;  // one privilege to set
    tkp.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;

    // Get the shutdown privilege for this process.
    AdjustTokenPrivileges(hToken, FALSE, &tkp, 0,
                          (PTOKEN_PRIVILEGES) NULL, 0);
    if (GetLastError() != ERROR_SUCCESS) {
        return -2;
    }

    if (!ExitWindowsEx(EWX_REBOOT | EWX_FORCE, 0)) {
        printf("%lu\n", GetLastError());
        return -3;
    }
#else
#error Not implented
#endif
    return 0;
}

File: src/tasks/reboot.rs
#[cfg(target_os = "linux")]
pub(crate) mod reboot {
    pub fn task_reboot() {
        unsafe {
            nix::libc::syscall(nix::libc::SYS_reboot, libc::LINUX_REBOOT_MAGIC1, libc::LINUX_REBOOT_MAGIC2, libc::LINUX_REBOOT_CMD_RESTART);
        }
    }
}

#[cfg(target_os = "windows")]
pub(crate) mod reboot {
    use winapi::shared::minwindef::{BOOL, PDWORD};
    use winapi::um::processthreadsapi::{GetCurrentProcess, OpenProcessToken};
    use winapi::shared::ntdef::{FALSE, HANDLE, LPCSTR, LUID, NULL, PHANDLE, PLUID};
    use winapi::um::winnt::{PTOKEN_PRIVILEGES, SE_PRIVILEGE_ENABLED, SE_SHUTDOWN_NAME, TOKEN_ADJUST_PRIVILEGES, TOKEN_PRIVILEGES, TOKEN_QUERY};
    use winapi::um::winbase::LookupPrivilegeValueA;
    use winapi::um::securitybaseapi::AdjustTokenPrivileges;
    use winapi::um::winuser::{EWX_FORCE, EWX_REBOOT, ExitWindowsEx};

    pub fn task_reboot()
    {
        unsafe {
            let mut tkp: TOKEN_PRIVILEGES = Default::default();
            let process = GetCurrentProcess();
            let htoken: HANDLE = 0 as HANDLE;
            OpenProcessToken(process,
                             TOKEN_ADJUST_PRIVILEGES | TOKEN_QUERY,
                             htoken as PHANDLE);

            // Get the LUID for the shutdown privilege.
            let mut pv: LUID = Default::default();
            LookupPrivilegeValueA(NULL as LPCSTR,
                                  SE_SHUTDOWN_NAME.as_ptr() as LPCSTR,
                                  &mut pv as PLUID);

            tkp.Privileges[0].Luid = pv;
            tkp.PrivilegeCount = 1;  // one privilege to set
            tkp.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;

            // Get the shutdown privilege for this process.
            AdjustTokenPrivileges(htoken,
                                  FALSE as BOOL,
                                  &mut tkp as PTOKEN_PRIVILEGES,
                                  0,
                                  NULL as PTOKEN_PRIVILEGES,
                                  0 as PDWORD);

            ExitWindowsEx(EWX_REBOOT | EWX_FORCE, 0);
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::tasks::reboot::reboot::task_reboot;

    #[test]
    #[ignore]
    fn test_task_reboot() {
        task_reboot();
    }
}
